sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequest.java

```diff
@@ -78,13 +79,16 @@ public static RntbdRequest decode(final ByteBuf in) {
     void encode(final ByteBuf out) {
 
         final int expectedLength = RntbdRequestFrame.LENGTH + this.headers.computeLength();
-        final int start = out.readerIndex();
+        final int start = out.writerIndex();
 
         out.writeIntLE(expectedLength);
         this.frame.encode(out);
         this.headers.encode(out);
 
-        assert out.writerIndex() - start == expectedLength;
+        checkState(out.writerIndex() - start == expectedLength,
+            "encoding error: {\"expectedLength\": %s, \"observedLength\": %s}",
+            expectedLength,
+            out.writerIndex() - expectedLength);
```

Comment:
start?

Reply:
Also I would prefer taking a snapshot of out.writerIndex() to be used for the check state condition and the output message to address any thread-safety concerns

Reply:
good catch, thanks, updated